### PR TITLE
[Docs] [Builtins] Add a comment about 'mkMachineParameters' failing to inline

### PR DESCRIPTION
**Pull request by effectfully**
:cry: 

Also, apparently I've practically predicted that:

Screenshot from 2022-08-03 20-00-01

### CODE REVIEW

**Comment by effectfully** [comment]
Sorry, don't review yet, need to fix the wording.

**Comment by effectfully** [comment]
Ok, ready for review.

**Comment by effectfully** [comment]
For reproducing:

1. checkout into the latest master
2. add `{-# OPTIONS_GHC -ddump-simpl -ddump-to-file -dsuppress-all -fforce-recomp -dumpdir /tmp/dumps #-}` to `PlutusCore.Evaluation.Machine.MachineParameters.ImmediateMachineParameters`
3. run `cabal build plutus-core:lib:plutus-core` and let it finish
4. go to `/tmp/dumps/ImmediateMachineParameters.dump-simpl`, observe thousands of lines of GHC Core
5. go to `PlutusCore.Evaluation.Machine.MachineParameters`
change the type signature of `mkMachineParameters` to
```mkMachineParameters ::
    ( HasMeaningIn uni (val uni fun)
    , ToBuiltinMeaning uni fun
    )
    => UnliftingMode
    -> CostModel machinecosts (CostingPart uni fun)
    -> MachineParameters machinecosts val uni fun```
6. run `cabal build plutus-core:lib:plutus-core` and let it finish
7. go to `/tmp/dumps/ImmediateMachineParameters.dump-simpl`, observe that there's only a few dozens of lines of GHC core there now

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs [hunk 1]
```diff
@@ -48,7 +48,9 @@ data MachineParameters machinecosts term (uni :: Type -> Type) (fun :: Type) =
 -- See Note [Inlining meanings of builtins].
 {-| This just uses 'toBuiltinsRuntime' function to convert a BuiltinCostModel to a BuiltinsRuntime. -}
 mkMachineParameters ::
-    ( -- In Cek.Internal we have `type instance UniOf (CekValue uni fun) = uni`, but we don't know that here.
+    ( -- WARNING: do not discharge the equality constraint as that causes GHC to fail to inline the
+      -- /caller/ of this function at /its/ call site. I.e. 'mkMachineParameters' continues to
+      -- inline, but its caller doesn't. Because of the lack of the equality constraint. Go figure.
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs:53]: Can you put the steps to reproduce (or a link to it) here?

**Review by zliu41 — APPROVED** [review]

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs [hunk 2]
```diff
@@ -48,7 +48,9 @@ data MachineParameters machinecosts term (uni :: Type -> Type) (fun :: Type) =
 -- See Note [Inlining meanings of builtins].
 {-| This just uses 'toBuiltinsRuntime' function to convert a BuiltinCostModel to a BuiltinsRuntime. -}
 mkMachineParameters ::
-    ( -- In Cek.Internal we have `type instance UniOf (CekValue uni fun) = uni`, but we don't know that here.
+    ( -- WARNING: do not discharge the equality constraint as that causes GHC to fail to inline the
+      -- /caller/ of this function at /its/ call site. I.e. 'mkMachineParameters' continues to
```
Comment (zliu41) [inline plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs:52]: Actually it _is_ `mkMachineParameters` that fails to be inlined. In the core dump I'm seeing
```
immediateMachineParameters1
  = $smkMachineParameters UnliftingImmediate
```

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs:52]: Oh indeed, I got confused. Thank you, will fix.

File: plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs (hunk 1 above)
Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs:53]: Will do.

**Comment by effectfully** [comment]
@zliu41 ready for another review. Thank you for figuring it out.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**